File: 2. Structural Design Patterns/Composite/Go/main.go
package main

import "fmt"

// Composite Pattern Example in Go
// File system: File (leaf) and Directory (composite)

type Component interface{ Display(int); GetSize() int }

type File struct{ name string; size int }
func (f File) Display(depth int){ fmt.Printf("%sFile: %s (%dKB)\n", indent(depth), f.name, f.size) }
func (f File) GetSize() int{ return f.size }

type Directory struct{ name string; items []Component }
func (d *Directory) Add(c Component){ d.items = append(d.items, c) }
func (d Directory) Display(depth int){
	fmt.Printf("%sDirectory: %s/\n", indent(depth), d.name)
	for _, item := range d.items{ item.Display(depth+1) }
}
func (d Directory) GetSize() int{
	total := 0
	for _, item := range d.items{ total += item.GetSize() }
	return total
}

func indent(depth int) string{ s := ""; for i:=0; i<depth*2; i++{ s += " " }; return s }

func main(){
	root := &Directory{name: "root"}
	docs := &Directory{name: "docs"}
	docs.Add(File{name: "readme.txt", size: 5})
	docs.Add(File{name: "manual.pdf", size: 150})
	root.Add(docs)
	root.Add(File{name: "config.ini", size: 2})
	root.Display(0)
	fmt.Printf("Total size: %dKB\n", root.GetSize())
}

File: 2. Structural Design Patterns/Composite/Rust/main.rs
// Composite Pattern Example in Rust
// File system: File (leaf) and Directory (composite)

trait FSItem {
    fn display(&self, depth: usize);
    fn size(&self) -> usize;
}

struct FSFile { name: String, size: usize }
impl FSItem for FSFile {
    fn display(&self, depth: usize) {
        println!("{}File: {} ({}KB)", " ".repeat(depth*2), self.name, self.size);
    }
    fn size(&self) -> usize { self.size }
}

struct Directory { name: String, items: Vec<Box<dyn FSItem>> }
impl Directory { fn add(&mut self, item: Box<dyn FSItem>) { self.items.push(item) } }
impl FSItem for Directory {
    fn display(&self, depth: usize) {
        println!("{}Directory: {}/", " ".repeat(depth*2), self.name);
        for it in &self.items { it.display(depth+1); }
    }
    fn size(&self) -> usize { self.items.iter().map(|i| i.size()).sum() }
}

fn main(){
    let mut root = Directory { name: "root".into(), items: vec![] };
    let mut docs = Directory { name: "docs".into(), items: vec![] };
    docs.add(Box::new(FSFile { name: "readme.txt".into(), size: 5 }));
    docs.add(Box::new(FSFile { name: "manual.pdf".into(), size: 150 }));
    root.add(Box::new(docs));
    root.add(Box::new(FSFile { name: "config.ini".into(), size: 2 }));
    root.display(0);
    println!("Total size: {}KB", root.size());
}
